### ISSUE TITLE
Implement the ability to search OpenSea Collectables when creating a new permission

### ISSUE THREAD

**Issue by John-44** [status-im/status-desktop#11462]
See designs here: https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?type=design&node-id=23289%3A500766&mode=design&t=qSxLPn009YqUtemm-1

When creating or editing permission, it should be possible for the user to reference any collectable listed on OpenSea in the permission.

![2023-07-08_16-27-53](https://github.com/status-im/status-desktop/assets/22846356/28e16d48-89aa-48ca-9917-152f240d87bb)

cc'ing @jrainville @iurimatias

**Comment by John-44** [comment]
@micieslak just speaking to @dlipicar , we might need to Infura or Alchamy API for this.  @dlipicar can help/do/advise on the backend here

**Comment by John-44** [comment]
@benjthayer depending on what @dlipicar discovers is possible via whatever NFT search API(s) we choose to use, we might need to slightly update this design.  Nothing to do for now though, the API research needs to be completed first.

**Comment by dlipicar** [comment]
Initial findings:

If we have some certainty that what the user entered is some specific piece of information (contract address, collection slug), we can make a search using specific endpoints in any of our providers. For example:
https://docs.opensea.io/reference/retrieve-nfts-by-contract
https://docs.opensea.io/reference/retrieve-nfts-by-collection
https://docs.infura.io/infura-expansion-apis/nft-api/rest-apis/api-reference/collections#get-all-nfts-with-metadata-in-a-specific-collection
https://docs.alchemy.com/reference/getnftsforcontract-v3

If we want to throw whatever the user enters to an API Endpoint and see if it matches any some field in the NFT metadata (For example enter "bored" and get the Bored Apes collection)
- OpenSea does not seem to provide such an API Endpoint.
- Infura does: https://docs.infura.io/infura-expansion-apis/nft-api/rest-apis/api-reference/metadata#search-for-nfts-given-a-string. Supports Ethereum Mainnet, Ethereum Testnet and Arbitrum Mainnet, see https://docs.infura.io/infura-expansion-apis/nft-api/rest-apis/supported-networks.
- Alchemy does: https://docs.alchemy.com/reference/searchcontractmetadata-v3. Supports only Ethereum Mainnet. We'll possibly want to move to a higher-priced plan if we start using this, need to check the cost of this endpoint.
- Moralis does (https://docs.moralis.io/web3-data-api/evm/reference/search-nfts?chain=eth&format=decimal&filter=name&addresses=[]), but the endpoint will be deprecated soon and it seems there will be no replacement.

So, we could get contract address/collection slug search support using
-OpenSea: All chains
-Infura: Eth Main/Test + Arb Main
And text-based search support using
-Infura: Eth Main/Test + Arb Main
-Alchemy: Eth Main

Probably a good idea to make some tests before making any decision.

**Comment by John-44** [comment]
thanks for looking at this @dlipicar 

Agree that it would be a good idea to do some tests before making any decision.

For 'Contract address/collection slug search' using either: OpenSea with Infura as a fallback OR just Infura, both work from a product POV.

For 'text-based search', using either: Infura with Alchemy as a fallback OR just Infura, both work from a product POV.

**Comment by dlipicar** [comment]
Update: Infura's shutting down their NFT API
![image](https://github.com/status-im/status-desktop/assets/11161531/7b8e71a3-cd7b-4007-83fb-08d3da31343c)

**Comment by John-44** [comment]
@dlipicar @alaibe Can we use Moralis instead as recommended by infura in their message above?

**Comment by dlipicar** [comment]
@John-44 I wouldn't recommend [Moralis](https://docs.moralis.io/web3-data-api/evm/reference/search-nfts?chain=eth&format=decimal&filter=name&addresses=[]) for the string-based search feature: 

![image](https://github.com/status-im/status-desktop/assets/11161531/d17355c7-bb8e-4ff2-a340-b7704652315e)

I'd go with [Alchemy](https://docs.alchemy.com/reference/searchcontractmetadata-v3) to begin with, need to look for a suitable backup provider.

**Comment by John-44** [comment]
> @John-44 I wouldn't recommend [Moralis](https://docs.moralis.io/web3-data-api/evm/reference/search-nfts?chain=eth&format=decimal&filter=name&addresses=%5B%5D) for the string-based search feature:
> 
> ![image](https://user-images.githubusercontent.com/11161531/262480805-d17355c7-bb8e-4ff2-a340-b7704652315e.png)
> 
> I'd go with [Alchemy](https://docs.alchemy.com/reference/searchcontractmetadata-v3) to begin with, need to look for a suitable backup provider.

Sounds good, let me know if I need to do anything in order for us to use Alchemy (e.g. if we have to pay them).

**Comment by dlipicar** [comment]
Additional note: Rarible seems to have something we could use, though they only support Ethereum (mainnet + testnet)
https://multichain.redoc.ly/testnet/tag/item-controller#operation/searchItems

**Comment by alaibe** [comment]
Additional note:
First search with alchemy: https://docs.alchemy.com/reference/searchcontractmetadata-v3
Then Rarible: https://docs.rarible.org/reference/searchitems

**Comment by dlipicar** [comment]
Ugh... Rarible's API has a case-sensitive text field, so searching for `lilpudgys` returns a completely different list than `Lilpudgys` 🤦 
Asked their support if we can get case-sensitivity disabled or at least configurable

**Comment by dlipicar** [comment]
Iterating over the designs with Ben

**Comment by benjthayer** [comment]
Iterated on with Dario and pending review with @John-44 and design team...
https://www.figma.com/file/qHfFm7C9LwtXpfdbxssCK3/Kuba%E2%8E%9CDesktop---Communities?type=design&node-id=53449-148202&mode=design&t=sYCpMDPq54tgNmF8-4

cc @dlipicar

**Comment by dlipicar** [comment]
I'm guessing the priority for this is now quite low, won't resume work on it until further notice

**Comment by John-44** [comment]
> I'm guessing the priority for this is now quite low, won't resume work on it until further notice

We do need a way for users to create permissions tied to any arbitrary NFT, not just NFTs minted in status.  How much work is remaining to complete the implementation of this, could anybody else pick this up?

**Comment by dlipicar** [comment]
@John-44 oh sorry, I thought this was considered a "new messenger feature" and had been de-prioritized. My bad!
 
The backend is mostly done, the UI is mostly not. I can finish the few missing pieces from the backend and see if somebody from the UI team can work on the rest. Do we want this in for 2.29?

**Comment by John-44** [comment]
> Do we want this in for 2.29?

If possible, yes 
cc'ing @noeliaSD

**Comment by benjthayer** [comment]
@dlipicar designs are now signed off an folded into main permissions flows in Figma :)
https://www.figma.com/file/qHfFm7C9LwtXpfdbxssCK3/Kuba%E2%8E%9CDesktop---Communities?type=design&node-id=53874-101256&mode=design&t=dVuLDOqra39yKahC-4

**Comment by noeliaSD** [comment]
> https://www.figma.com/file/qHfFm7C9LwtXpfdbxssCK3/Kuba%E2%8E%9CDesktop---Communities?type=design&node-id=53874-101256&mode=design&t=dVuLDOqra39yKahC-4



> > Do we want this in for 2.29?
> 
> If possible, yes cc'ing @noeliaSD

Hey @John-44 we cannot make it for `2.29` but we can work on it for `2.30`. Is it ok?

**Comment by John-44** [comment]
> Hey @John-44 we cannot make it for `2.29` but we can work on it for `2.30`. Is it ok?

This is fine, thanks for the heads up :-)
